woocommerce/woocommerce-ios
# Create wrappers for `ProductFormDataModel` conformance based on `Product` and `ProductVariation`

Fixes #2575 

I apologized for the large PR, most of the diffs are just from renaming, and all the diffs in unit tests (probably more than half of the total diffs) are just creating the wrapper to replace in the original tests without new logic.

## Changes

- Created `EditableProductModel` and `EditableProductVariationModel` as wrappers that conform to `ProductFormDataModel` based on `Product` and `ProductVariation` respectively.
  - The goal here is so that `Product` and `ProductVariation` don't directly implement `ProductFormDataModel` protocol, since a few of the methods are feature-specific and it doesn't make sense for a `ProductVariation` to have a `shortDescription` just because it has to be used in a generic product form (original discussion https://github.com/woocommerce/woocommerce-ios/pull/2566#discussion_r462664774). Another goal is so that if the data model has dependencies other than `Product`/`ProductVariation`, they can be injected to the wrapper. An example use case is to support the variation name update in https://github.com/woocommerce/woocommerce-ios/issues/2595 that requires its parent product's attributes.
- Replaced all the existing concrete usage of `Product` and `ProductVariation` in product form related classes with `EditableProductModel` and `EditableProductVariationModel`
  - I meant to rename `product` to `productModel` or `model` in product form and a few related view controllers, but thought this PR is already very large. After the renaming, it would be less awkward to see some `product.product` calls to access `Product` for `Product` specific operations
- In `ProductFormDataModel`, replaced `virtual` and `downloadable` properties with `func isShippingEnabled() -> Bool` since these two properties were only used for this shipping logic
- Renamed a few `Product`/`ProductVariation` checks to `EditableProductModel`/`EditableProductVariationModel` respectively.
- Updated unit tests (this was the most time consuming part!)

## Testing

No app behavior changes are expected, please feel free to sanity check the product and variation form UI:

- Go to the Products tab
- Tap on a simple product --> make sure all the rows and bottom sheet actions navigate to the expected screens, and product settings are accessible
- Feel free to edit a few properties and update --> the product should be saved remotely
- Go back to the Products tab
- Tap on a variable product
- Tap on the variations row
- Tap on a variation --> make sure all the rows and bottom sheet actions navigate to the expected screens
- Feel free to edit a few properties and update --> the variation should be saved remotely

Update release notes:
